src/lib/math/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,1344 @@
+
+      using AffinePoint = AffineCurvePoint<FieldElement, Params>;
+      using ProjectivePoint = ProjectiveCurvePoint<FieldElement, Params>;
+
+      static const constinit size_t OrderBits = Scalar::BITS;
+      static const constinit size_t PrimeFieldBits = FieldElement::BITS;
+
+      static constexpr FieldElement A = FieldElement::from_words(Params::AW);
+      static constexpr FieldElement B = FieldElement::from_words(Params::BW);
+
+      static constexpr AffinePoint G =
+         AffinePoint(FieldElement::from_words(Params::GXW), FieldElement::from_words(Params::GYW));
+
+      static constexpr FieldElement SSWU_Z = FieldElement::constant(Params::Z);
+
+      static const constinit bool ValidForSswuHash =
+         (SSWU_Z.is_nonzero() && A.is_nonzero() && B.is_nonzero() && FieldElement::P_MOD_4 == 3);
+
+      // (-B / A), will be zero if A == 0 or B == 0 or Z == 0
+      static const FieldElement& SSWU_C1()
+         requires ValidForSswuHash
+      {
+         // We derive it from C2 to avoid a second inversion
+         static const auto C1 = (SSWU_C2() * SSWU_Z).negate();
+         return C1;
+      }
+
+      // (B / (Z * A)), will be zero if A == 0 or B == 0 or Z == 0
+      static const FieldElement& SSWU_C2()
+         requires ValidForSswuHash
+      {
+         // This could use a variable time inversion
+         static const auto C2 = (B * (SSWU_Z * A).invert());
+         return C2;
+      }
+};
+
+template <typename C, size_t WindowBits>
+class BlindedScalarBits final {
+   private:
+      typedef typename C::W W;
+
+      // Use 1/3 the order, rounded up to the next word for blinding
+      static const constinit size_t BlindingBits =
+         ((C::OrderBits / 3 + WordInfo<W>::bits - 1) / WordInfo<W>::bits) * WordInfo<W>::bits;
+
+      static_assert(BlindingBits % WordInfo<W>::bits == 0);
+      static_assert(BlindingBits < C::Scalar::BITS);
+
+   public:
+      static constexpr size_t Bits = C::Scalar::BITS + BlindingBits;
+
+      BlindedScalarBits(const typename C::Scalar& scalar, RandomNumberGenerator& rng) {
+         if constexpr(BlindingBits > 0) {
+            constexpr size_t mask_words = BlindingBits / WordInfo<W>::bits;
+            constexpr size_t mask_bytes = mask_words * WordInfo<W>::bytes;
+
+            constexpr size_t n_words = C::NW.size();
+
+            uint8_t maskb[mask_bytes] = {0};
+            rng.randomize(maskb, mask_bytes);
+
+            W mask[n_words] = {0};
+            load_le(mask, maskb, mask_words);
+            mask[mask_words - 1] |= WordInfo<W>::top_bit;
+
+            W mask_n[2 * n_words] = {0};
+
+            const auto sw = scalar.to_words();
+
+            // Compute masked scalar s + k*n
+            comba_mul<n_words>(mask_n, mask, C::NW.data());
+            bigint_add2_nc(mask_n, 2 * n_words, sw.data(), sw.size());
+
+            std::reverse(mask_n, mask_n + 2 * n_words);
+            m_bytes = store_be<secure_vector<uint8_t>>(mask_n);
```

Comment:
This would become more compact if `comba_mul` and `bigint_add2_nc` would be able to handle spans.

```suggestion
            constexpr size_t mask_words = BlindingBits / WordInfo<W>::bits;
            constexpr size_t n_words = C::NW.size();

            std::array<W, n_words> mask = {};
            std::array<W, 2 * n_words> masked_scalar = {};

            auto effective_mask = std::span{mask}.template first<mask_words>();
            load_le(effective_mask, rng.random_array<effective_mask.size_bytes()>());
            effective_mask.back() |= WordInfo<W>::top_bit;

            const auto sw = scalar.to_words();

            // Compute masked scalar s + k*n
            comba_mul<n_words>(masked_scalar.data(), mask.data(), C::NW.data());
            bigint_add2_nc(masked_scalar.data(), masked_scalar.size(), sw.data(), sw.size());

            std::reverse(masked_scalar.begin(), masked_scalar.end());
            m_bytes = store_be<secure_vector<uint8_t>>(masked_scalar);
```

... also, this assumes that we add the following overload to `RandomNumberGenerator`:

```C++
/**
* Create an array of @p N random bytes.
*
* @tparam N  number of desired random bytes
* @return    an array of @p N random bytes
* @throws Exception if RNG fails
*/
template <size_t N>
std::array<uint8_t, N> random_array() {
   std::array<uint8_t, N> result;
   random_vec(std::span{result});
   return result;
}
```